skel/file/mod/lib.mustache

```diff
@@ -0,0 +1,424 @@
+{{!
+    lib.php
+
+    * component
+    * copyright
+}}
+{{< common/lib }}
+{{/ common/lib }}
+
+/**
+ * Return if the plugin supports $feature.
+ *
+ * @param string $feature Constant representing the feature.
+ * @return true | null True if the feature is supported, null otherwise.
+ */
+function {{ component }}_supports($feature) {
+    switch ($feature) {
+        {{# self.supports }}
+        case {{ . }}:
+            return true;
+        {{/ self.supports }}
+        default:
+            return null;
+    }
+}
+
+/**
+ * Saves a new instance of the {{ component }} into the database.
+ *
+ * Given an object containing all the necessary data, (defined by the form
+ * in mod_form.php) this function will create a new instance and return the id
+ * number of the instance.
+ *
+ * @param object $moduleinstance An object from the form.
+ * @param mod_{{ component_name }}_mod_form $mform The form.
+ * @return int The id of the newly inserted record.
+ */
+function {{ component }}_add_instance($moduleinstance, $mform = null) {
+    global $DB;
+
+    $moduleinstance->timecreated = time();
+
+    $id = $DB->insert_record('{{ component }}', $moduleinstance);
+
+    return $id;
+}
+
+/**
+ * Updates an instance of the {{ component }} in the database.
+ *
+ * Given an object containing all the necessary data (defined in mod_form.php),
+ * this function will update an existing instance with new data.
+ *
+ * @param object $moduleinstance An object from the form in mod_form.php.
+ * @param mod_{{ component_name }}_mod_form $mform The form.
+ * @return bool True if successful, false otherwise.
+ */
+function {{ component }}_update_instance($moduleinstance, $mform = null) {
+    global $DB;
+
+    $moduleinstance->timemodified = time();
+    $moduleinstance->id = $moduleinstance->instance;
+
+    return $DB->update_record('{{ component }}', $moduleinstance);
+}
+
+/**
+ * Removes an instance of the {{ component }} from the database.
+ *
+ * @param int $id Id of the module instance.
+ * @return bool True if successful, false on failure.
+ */
+function {{ component }}_delete_instance($id) {
+    global $DB;
+
+    $exists = $DB->get_record('{{ component }}', array('id' => $id));
+    if (!$exists) {
+        return false;
+    }
+
+    $DB->delete_records('{{ component }}', array('id' => $id));
+
+    return true;
+}
+{{# self.functions.user_outline }}
```

Comment:
Another option for generating feature specific functions would be to have a check like {{# self.has_gradebook }} instead of checking for each individual function (like I did for navigation).